js/private/js_image_layer.bzl

```diff
@@ -0,0 +1,226 @@
+    real_executable_path = _runfile_path(ctx, executable, runfiles_dir)
+    launcher = _write_laucher(ctx, real_executable_path)
+
+    files = {}
+
+    files[executable_path] = {"dest": launcher.path, "root": launcher.root.path}
+
+    for file in depset(transitive = [default_info.files, default_info.default_runfiles.files]).to_list():
+        destination = _runfile_path(ctx, file, runfiles_dir)
+        entry = {"dest": file.path, "root": file.root.path, "is_source": file.is_source, "is_directory": file.is_directory}
+        if destination == real_executable_path:
+            entry["remove_non_hermetic_lines"] = True
+        files[destination] = entry
+
+    entries = ctx.actions.declare_file("{}_entries.json".format(ctx.label.name))
+    ctx.actions.write(entries, content = json.encode(files))
+
+    extension = ".tar"
+
+    if ctx.attr.compression == "gzip":
+        extension = ".tar.gz"
+
+    app = ctx.actions.declare_file("{name}_app{extension}".format(name = ctx.label.name, extension = extension))
+    node_modules = ctx.actions.declare_file("{name}_node_modules{extension}".format(name = ctx.label.name, extension = extension))
+
+    args = ctx.actions.args()
+    args.add(entries)
+    args.add(app)
+    args.add(node_modules)
+    args.add(ctx.attr.compression if ctx.attr.compression else "none")
+
+    ctx.actions.run(
+        inputs = depset([executable, launcher, entries], transitive = [default_info.files, default_info.default_runfiles.files]),
+        outputs = [app, node_modules],
+        arguments = [args],
+        executable = ctx.executable._builder,
+        progress_message = "JsImageLayer %{label}",
+        env = {
+            "BAZEL_BINDIR": ".",
+        },
+    )
+
+    return [
+        DefaultInfo(files = depset([app, node_modules])),
+        OutputGroupInfo(app = depset([app]), node_modules = depset([node_modules])),
+    ]
+
+def _js_image_layer_transition_impl(settings, attr):
+    # buildifier: disable=unused-variable
+    _ignore = (settings)
+    if not attr.platform:
+        return {}
+    return {
+        "//command_line_option:platforms": str(attr.platform),
+    }
+
+_js_image_layer_transition = transition(
+    implementation = _js_image_layer_transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:platforms"],
+)
+
+js_image_layer = rule(
+    implementation = _js_image_layer_impl,
+    doc = _doc,
+    attrs = {
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+        "binary": attr.label(mandatory = True, cfg = _js_image_layer_transition, doc = "Label to an js_binary target"),
+        "root": attr.string(doc = "Path where the files from js_binary will reside in. eg: /apps/app1 or /app"),
+        "compression": attr.string(doc = "Compression algorithm. Can be one of `gzip`, `none`.", values = ["gzip", "none"], default = "gzip"),
```

Comment:
nit: is this buildifier-formatted? make sure you've got the `pre-commit install`ed

I expected to see more newlines/indents in these attribute lists